### PR TITLE
DUOS-1642[risk=no] UI dataset required sub section in data use bucket

### PR DESCRIPTION
**Pull request by shaemarks**
[Link to Jira ticket](https://broadworkbench.atlassian.net/browse/DUOS-1642)
Summary of changes: 

- Filter datasets to only include those that belong to both the given bucket and given list of datasets for this dac
- When collapsed, display the first 5 datasets and a link to expand
- When expanded, display all datasets and a link to collapse 

----
Have you read [Terra's Contributing Guide](https://github.com/DataBiosphere/terra-ui/wiki/Contributor-Guide) lately? If not, do that first.

- Label PR with a Jira ticket number and include a link to the ticket
- Label PR with a security risk modifier [no, low, medium, high]
- PR describes scope of changes
- Get a minimum of one thumbs worth of review, preferably two if enough team members are available
- Get PO sign-off for all non-trivial UI or workflow changes
- Verify all tests go green
- Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

**Comment by shaemarks** [comment]
To test locally, I added the code below to `DarCollectionReview.js`  and went to `http://localhost:3000/dar_collection/{any collection ID}`. The dataSetIds listed under `bucket.elections` correspond to the datasets for the Dac with dacID 3 (should be displayed), but other ids can be added for when the bucket and dar datasets do not match.

```
 const [dacDatasets, setDacDatasets] = useState([]);


 const dacDatasets = await DAC.datasets(3);
 setDacDatasets(dacDatasets);


   h(DatasetsRequestedPanel, {
        bucket: {
          elections: [[
            {dataSetId: 1398},
            {dataSetId: 1399},
            {dataSetId: 459},
            {dataSetId: 1403},
            {dataSetId: 384},
            {dataSetId: 1396},
            {dataSetId: 1397},
            {dataSetId: 390},
          ]]
        },
        dacDatasets,
        isLoading
      })
```

File: cypress/component/ResearchProposalVoteSlab/datasets_requested.spec.js [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+
+const dataset = (id) => {
+  return {
+    dataSetId: id,
+    alias: `DUOS-${id}`,
+    properties: [
+      {propertyName: "Dataset Name", propertyValue: `Dataset ${id}`}
+    ]
+  }
+}
```
Comment (rushtong) [inline cypress/component/ResearchProposalVoteSlab/datasets_requested.spec.js]: nit: missing semi-colons

File: cypress/component/ResearchProposalVoteSlab/datasets_requested.spec.js [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+
+const dacDatasets = [
+  dataset(1),
+  dataset(2),
+  dataset(3),
+  dataset(4),
+  dataset(5),
+  dataset(6),
+  dataset(7),
+]
```
Comment (rushtong) [inline cypress/component/ResearchProposalVoteSlab/datasets_requested.spec.js]: nit: same here

File: src/components/collection_voting_slab/DatasetsRequestedPanel.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+  };
+
+  const datasetName = (dataset) => {
+    const datasetNameProperty = !isNil(dataset.properties) &&
+      find(dataset.properties, property => {
+        return property.propertyName === 'Dataset Name';
+      });
+
+    return datasetNameProperty ? datasetNameProperty.propertyValue : '- -';
+  };
```
Comment (rushtong) [inline src/components/collection_voting_slab/DatasetsRequestedPanel.js]: The dataset's name is just `dataset.name`

File: src/components/collection_voting_slab/DatasetsRequestedPanel.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+      ]);
+    });
+    return isLoading
+      ? div({className: 'text-placeholder', style: styles.skeletonLoader})
+      : div({style: styles.datasetList, dataCy: 'dataset-list'}, [datasetRows]);
+  };
+
+  const datasetId = (dataset) => {
+    return !isNil(dataset.alias) ? dataset.alias : '- -';
+  };
```
Comment (rushtong) [inline src/components/collection_voting_slab/DatasetsRequestedPanel.js]: The ID we show on the catalog page is labeled `dataset.datasetIdentifier` in the returned dataset we have in the DAR Collection

**Review by rushtong — CHANGES_REQUESTED** [review]
Very minor comments, but otherwise, code looks very good!

File: src/components/collection_voting_slab/DatasetsRequestedPanel.js (hunk 3 above)
Comment (shaemarks) [reply src/components/collection_voting_slab/DatasetsRequestedPanel.js]: I see `dataset.name` and `dataset.datasetIdentifier` in the datasets within a whole collection, but I'm not seeing the same properties within the datasets retrieved from the `/api/dac/{dacId}/datasets` endpoint or the `processedBuckets`

Comment (rushtong) [reply src/components/collection_voting_slab/DatasetsRequestedPanel.js]: In both cases, we should have access to the database id of the dataset - is that not enough to do a comparison? Do we need to add anything to either endpoint to accomplish what you're trying to do?

Comment (shaemarks) [reply src/components/collection_voting_slab/DatasetsRequestedPanel.js]: We can compare the dataset ids from the endpoint and `processedBuckets` to filter which datasets we need, and then compare those matching dataset ids against the collection datasets to get the needed properties. The other option would be to alter  `/api/dac/{dacId}/datasets` to include those properties.

**Review by quazi-h — APPROVED** [review]
The new component code for a Datasets Requested sub-section looks good to me

File: src/components/collection_voting_slab/DatasetsRequestedPanel.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  const {bucketDatasetIds, dacDatasetIds, collectionDatasets, isLoading} = props;
+
+
+  const requestedDatasetIds = useCallback(() => {
+    const datasetsForDacInBucket =  filter(bucketDatasetId => {
+      return includes(bucketDatasetId)(dacDatasetIds);
+    })(bucketDatasetIds);
+
+    return filter(dataset => {
+      return includes(dataset.datasetId)(datasetsForDacInBucket);
```
Comment (rushtong) [inline src/components/collection_voting_slab/DatasetsRequestedPanel.js]: The dataset id property on the dataset is weirdly capitalized due to history ... this line should be:
```suggestion
      return includes(dataset.dataSetId)(datasetsForDacInBucket);
```

**Review by rushtong — CHANGES_REQUESTED** [review]
Code looks good! One more small fix, then 👍🏽

File: src/components/collection_voting_slab/DatasetsRequestedPanel.js (hunk 5 above)
Comment (rushtong) [reply src/components/collection_voting_slab/DatasetsRequestedPanel.js]: Just an FYI for testing purposes, I added this component to the DarCollectionReview page and viewed collection id 190 on dev:
```
      h(DatasetsRequestedPanel, {
        bucketDatasetIds: [1250, 449],
        dacDatasetIds: [1250],
        collectionDatasets: collection.datasets,
        isLoading
      }),
```
Fiddling with the bucketDatasetIds and dacDatasetIds shows me the results I was expecting to see.

**Review by rushtong — APPROVED** [review]
Thanks for the changes!

**Pull request merged**